### PR TITLE
Store the rules of SCA checks in the Wazuh database

### PR DESCRIPTION
**Pull request by Skeptor**
This PR is the development for the issue: https://github.com/wazuh/wazuh/issues/2960

The commit https://github.com/wazuh/wazuh/commit/423e85a39de2c6ec2d549c6f87d3b0b9e720d426 adds the table `sca_check_rules` to the agent database. This table contains the id of the check, the type of rule and the rule itself. If a check has multiple rules each one  will be stored in a different row.

The commit https://github.com/wazuh/wazuh/commit/4901c55ed3c747da76d4c4df8729803393118834 adds the rules of the SCA check to the SCA event sent to the manager. This way we can use the rules in the `analysisd` decoder and send it to the Wazuh DB. Also it adds the funcionality to fill the `sca_check_rules` table.

The commit https://github.com/wazuh/wazuh/commit/6f335883dd544164af6480d5e18443697c0e5b25 adds to Wazuh DB the capability to delete an entry of `sca_check_rules` when it is deleted from `sca_check` as it is done with the table `sca_check_compliances`

This PR must:
- [x] Fill the table `sca_check_rules` of the agent database
- [x] Delete the entries of a check that has been removed from a policy file

Testing process:
- [x] Test agents with differents OS 
  - [x] Windows
  - [x] MAC
  - [x] Centos 7
  - [x] Ubuntu 18
- [x] Run the modified modules with Valgrind
  - [x] `wazuh-modulesd`
  - [x] `ossec-analysisd`
  - [x] `wazuh-db`
- [x] Integrity checks
  - [x] New check in the policy
  - [x] Delete a check from a policy

### CODE REVIEW

**Comment by Skeptor** [comment]
Memory Test
-----------
I've got the following results at the execution of `valgrind` over the wazuh daemons:

`wazuh.db`
--------------
```
==11954== LEAK SUMMARY:
==11954==    definitely lost: 0 bytes in 0 blocks
==11954==    indirectly lost: 0 bytes in 0 blocks
==11954==      possibly lost: 272 bytes in 1 blocks
==11954==    still reachable: 8,344 bytes in 2 blocks
==11954==         suppressed: 0 bytes in 0 blocks
==11954== Reachable blocks (those to which a pointer was found) are not shown.
==11954== To see them, rerun with: --leak-check=full --show-leak-kinds=all
```
`ossec-analysisd`
----------------------
```
==11966== LEAK SUMMARY:
==11966==    definitely lost: 164 bytes in 10 blocks
==11966==    indirectly lost: 0 bytes in 0 blocks
==11966==      possibly lost: 28,832 bytes in 106 blocks
==11966==    still reachable: 13,157,634 bytes in 59,634 blocks
==11966==         suppressed: 0 bytes in 0 blocks
==11966== Reachable blocks (those to which a pointer was found) are not shown.
==11966== To see them, rerun with: --leak-check=full --show-leak-kinds=all
```
Where the definitely lost blocks are related to the load of the rules and to Windows Eventchannel decoder
```
==11966== 12 bytes in 2 blocks are definitely lost in loss record 38 of 476
==11966==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11966==    by 0x5B1C9B9: strdup (strdup.c:42)
==11966==    by 0x16FE95: Read_Rules (rules-config.c:202)
==11966==    by 0x15DC9E: read_main_elements (config.c:92)
==11966==    by 0x15E8D5: ReadConfig (config.c:234)
==11966==    by 0x11373C: GlobalConf (config.c:99)
==11966==    by 0x13074B: main (analysisd.c:382)
==11966== 
==11966== 12 bytes in 2 blocks are definitely lost in loss record 39 of 476
==11966==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11966==    by 0x5B1C9B9: strdup (strdup.c:42)
==11966==    by 0x170392: Read_Rules (rules-config.c:232)
==11966==    by 0x15DC9E: read_main_elements (config.c:92)
==11966==    by 0x15E8D5: ReadConfig (config.c:234)
==11966==    by 0x11373C: GlobalConf (config.c:99)
==11966==    by 0x13074B: main (analysisd.c:382)
==11966== 
==11966== 22 bytes in 1 blocks are definitely lost in loss record 82 of 476
==11966==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11966==    by 0x5B1C9B9: strdup (strdup.c:42)
==11966==    by 0x16F7CC: Read_Rules (rules-config.c:171)
==11966==    by 0x15DC9E: read_main_elements (config.c:92)
==11966==    by 0x15E8D5: ReadConfig (config.c:234)
==11966==    by 0x11373C: GlobalConf (config.c:99)
==11966==    by 0x13074B: main (analysisd.c:382)
==11966== 
==11966== 24 bytes in 2 blocks are definitely lost in loss record 92 of 476
==11966==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11966==    by 0x5B1C9B9: strdup (strdup.c:42)
==11966==    by 0x1700B0: Read_Rules (rules-config.c:216)
==11966==    by 0x15DC9E: read_main_elements (config.c:92)
==11966==    by 0x15E8D5: ReadConfig (config.c:234)
==11966==    by 0x11373C: GlobalConf (config.c:99)
==11966==    by 0x13074B: main (analysisd.c:382)
==11966== 
==11966== 30 bytes in 2 blocks are definitely lost in loss record 96 of 476
==11966==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11966==    by 0x5B1C9B9: strdup (strdup.c:42)
==11966==    by 0x16FBB3: Read_Rules (rules-config.c:186)
==11966==    by 0x15DC9E: read_main_elements (config.c:92)
==11966==    by 0x15E8D5: ReadConfig (config.c:234)
==11966==    by 0x11373C: GlobalConf (config.c:99)
==11966==    by 0x13074B: main (analysisd.c:382)
==11966== 
==11966== 64 bytes in 1 blocks are definitely lost in loss record 137 of 476
==11966==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11966==    by 0x4F1706B: cJSON_New_Item (cJSON.c:160)
==11966==    by 0x4F1A567: cJSON_CreateObject (cJSON.c:2163)
==11966==    by 0x144CE1: DecodeWinevt (winevtchannel.c:83)
==11966==    by 0x134F4E: w_decode_winevt_thread (analysisd.c:2164)
==11966==    by 0x58676DA: start_thread (pthread_create.c:463)
==11966==    by 0x5BA088E: clone (clone.S:95)

```

`wazuh-modulesd`
------------------------
```
==17329== LEAK SUMMARY:
==17329==    definitely lost: 0 bytes in 0 blocks
==17329==    indirectly lost: 0 bytes in 0 blocks
==17329==      possibly lost: 6,641 bytes in 11 blocks
==17329==    still reachable: 619,621 bytes in 3,722 blocks
==17329==                       of which reachable via heuristic:
==17329==                         length64           : 223,896 bytes in 124 blocks
==17329==         suppressed: 0 bytes in 0 blocks
```

Integrity
----------
To check the integrity I removed the check 1500 from the policy `system_audit_ssh` and I check that the others rows persist. For example, I query for the rules 1500 and 1501
```
sqlite> select * from sca_check_rules where id_check ="1500";
1500|file|f:$sshd_file -> !r:^# && r:Port\.+22;
sqlite> select * from sca_check_rules where id_check ="1501";
1501|file|f:$sshd_file -> !r:^# && r:Protocol\.+1;
```

After deleting the check 1500 from the policy file I got as a result of the querys:

```
sqlite> select * from sca_check_rules where id_check ="1500";
sqlite> select * from sca_check_rules where id_check ="1501";
1501|file|f:$sshd_file -> !r:^# && r:Protocol\.+1;
```

Operative Systems
---------------------------
In the OS I tested I scan the default policies for each one. In the `ossec.log` we can see what policies has been scanned in each system:

Ubuntu 18
---------------
```
2019/04/11 12:13:08 sca: INFO: Starting evaluation of policy: 'cis_debian_linux_rcl.yml
2019/04/11 12:13:11 sca: INFO: Evaluation finished for policy 'cis_debian_linux_rcl.yml'.
2019/04/11 12:13:28 sca: INFO: Starting evaluation of policy: 'system_audit_rcl.yml
2019/04/11 12:13:31 sca: INFO: Evaluation finished for policy 'system_audit_rcl.yml'.
2019/04/11 12:13:48 sca: INFO: Starting evaluation of policy: 'system_audit_ssh.yml
2019/04/11 12:13:51 sca: INFO: Evaluation finished for policy 'system_audit_ssh.yml'.
2019/04/11 12:14:08 sca: INFO: Starting evaluation of policy: 'system_audit_pw.yml
2019/04/11 12:14:11 sca: INFO: Evaluation finished for policy 'system_audit_pw.yml'
```

Centos 7
------------
```
2019/04/11 10:12:06 sca: INFO: Starting evaluation of policy: 'cis_rhel7_linux_rcl.yml
2019/04/11 10:12:14 sca: INFO: Evaluation finished for policy 'cis_rhel7_linux_rcl.yml'.
2019/04/11 10:12:15 sca: INFO: Starting evaluation of policy: 'system_audit_rcl.yml
2019/04/11 10:12:18 sca: INFO: Evaluation finished for policy 'system_audit_rcl.yml'.
2019/04/11 10:12:19 sca: INFO: Starting evaluation of policy: 'system_audit_ssh.yml
2019/04/11 10:12:22 sca: INFO: Evaluation finished for policy 'system_audit_ssh.yml'.
```

Mac OS
-----------
```
2019/04/11 11:03:05 sca: INFO: Starting evaluation of policy: 'cis_apple_macOS_10.13.yml
2019/04/11 11:03:05 sca: INFO: Evaluation finished for policy 'cis_apple_macOS_10.13.yml'.
2019/04/11 11:03:06 sca: INFO: Starting evaluation of policy: 'system_audit_rcl_mac.yml
2019/04/11 11:03:09 sca: INFO: Evaluation finished for policy 'system_audit_rcl_mac.yml'.
2019/04/11 11:03:10 sca: INFO: Starting evaluation of policy: 'system_audit_ssh.yml
2019/04/11 11:03:13 sca: INFO: Evaluation finished for policy 'system_audit_ssh.yml'.
```
Windows
------------
```
2019/04/11 09:17:08 sca: INFO: Starting evaluation of policy: 'win_audit_rcl.yml
2019/04/11 09:17:12 sca: INFO: Evaluation finished for policy 'win_audit_rcl.yml'.
```

File: src/wazuh_modules/wm_sca.c [hunk 1]
```diff
@@ -824,8 +852,13 @@ static int wm_sca_do_scan(OSList *p_list,cJSON *profile_check,OSStore *vars,wm_s
                 nbuf = p_check->valuestring;
                 mdebug2("Rule is: %s",nbuf);
 
+                /* Make a copy of the rule */
+                char *rule_cp;
+                os_strdup(nbuf, rule_cp);
```
Comment (chemamartinez) [inline src/wazuh_modules/wm_sca.c]: It is necessary to copy the `nbuf` variable for processing the rules?

Comment (Skeptor) [reply src/wazuh_modules/wm_sca.c]: Yes, it is necessary becasue the function `wm_sca_get_value` crops the string received as argument, so the rule is lost.

Comment (chemamartinez) [reply src/wazuh_modules/wm_sca.c]: Ok, makes sense. Thank you!

**Pull request merged**